README.md
# @ddadaal/next-typed-api-routes

![cli](https://img.shields.io/npm/v/@ddadaal/next-typed-api-routes-runtime?label=runtime)
![cli](https://img.shields.io/npm/v/@ddadaal/next-typed-api-routes-cli?label=cli)

Write a `Schema` interface in your API route file, and you get

- route parameters validation and type completion
- typed API clients
- faster JSON response serialization

all at one command!

Migrate from v0.2? See the [migration guide](#migrate-from-v02)!

# Requirement for the target Next.js Project

1. Use [TypeScript](https://nextjs.org/docs/basic-features/typescript)
2. Set `"baseUrl": "."` in `tsconfig.json` `compilerOptions` part

```json
{
  "compilerOptions": {
    "baseUrl": "."
  }
}
```
3. All codes should live in `src` dir [next.js src directory](https://nextjs.org/docs/advanced-features/src-directory)

# Get Started

1. Install the packages in your Next.js TypeScript project

```bash
npm install --save @ddadaal/next-typed-api-routes-runtime
npm install -D @ddadaal/next-typed-api-routes-cli
```

2. Create a API route `src/pages/api/test.ts` with the following content

```ts
import { route } from "@ddadaal/next-typed-api-routes-runtime";

interface Value {
  articleId: number;
}

export interface TestApiSchema {
  method: "POST";

  query: {
    a?: string | number;
  };

  body: {
    test: string;
  };

  responses: {
    200: { test: string; }
    403: { message: string; }
  }
}

export default route<TestApiSchema>("TestApiSchema", async (req) => {
  const { a } = req.query;
  const { test } = req.body;

  return { 200: { test: test + "" + a } };
});
```

3. Run `npx ntar schema && npx ntar client`

`api-routes-schemas.json` will be generated at cwd. You should never edit it directly. The project cannot start without this file.

`src/apis/api.ts` will be generated at `src/apis`.

4. Import the `api` variable from `src/apis/api.ts` to use the client.

```ts
import { api } from "src/apis/api";

api.testApi({ query: {}, body: { test: "123" } })
  .httpError(403, ({ message }) => { console.log(403, message); })
  .then(({ test }) => { console.log(test); });
```

# Updating existing API Routes

To convert a normal API Routes to a type checked one, all you need to do is

1. Write a valid `Schema` interface
2. Wrap the handler function with `route` function, specify the name of Schema interface as the type argument and first argument

The Get Started part provides an example of a correctly formatted API route file. `ntar` will report errors if Incorrect route file is found.

Run `ntar schema` when the `Schema` interface is changed.

Run `ntar client` when the HTTP method or URL or the name of schema is changed.

# Schema Interface Specification

The shape of Schema interface is defined as follows:

```ts
// The name of the schema must end with "Schema"
// and must be unique across the whole project
interface TestSchema {
  // Required. Must be a valid CAPITALIZED string literal type of HTTP method (GET, POST, PATCH)
  method: "POST";

  // Optional. Define the path param and query (to align with next.js)
  query: {
    // Supports most type constructs
    property?: string | number | AnotherInterface | Pick<{ number: string }, "number">;
    
    /**
     * You can also use jsdoc for more specific limits
     * You can use keywords and values from JSON Schema Draft-07
     * @format email
     * @maxLength 50
     */
    anotherProperty: string;
  }

  // Optional. Define the body
  body: {
    // Supports most type constructs
    property?: string | number | AnotherInterface | Pick<{ number: string }, "number">;
  } 

  // Required. Define the responses
  responses: {
    // Key as response code. 
    // Only one [200, 300) response should exist. It will be considered by clients as the success response
    // If the success response has no payload, the status code must be 204.
    200: {
      // Supports most type constructs
      property?: string | number | AnotherInterface | Pick<{ number: string }, "number">;
    };

    // If the code has no payload, set the type to null
    404: null;
  }
}
```

# Tips

- All schemas and used models must have globally unique name
- Return a `{ [statusCode]: payload }` object in a route to take advantages of response body type check and faster JSON serialization
- To ensure that client bundle doesn't include unnecessary packages (e.g. `ajv`, `fast-json-stringify`, which are only used in server side), 
    - import packages only from `@ddadaal/next-typed-api-routes-runtime/lib/client` in client files, 
    - import types with `import type` clause

# Thanks

- [`Ajv`](https://ajv.js.org/) for JSON Schema validation
- [`vega/ts-json-schema-generator`](https://github.com/vega/ts-json-schema-generator) for generating json schema from typescript
- [`fast-json-stringify`](https://github.com/fastify/fast-json-stringify) for faster JSON response serialization
- [`fastify`](https://github.com/fastify/fastify) for inspiration or unified validation and serialization using JSON schema


# Migrate from v0.2

Since 0.3, this package is separated to [cli](packages/cli) and [runtime](packages/runtime) and organized as a monorepo. To migrate existing codebase, 

1. Delete previous dependency

```bash
npm uninstall --save @ddadaal/next-typed-api-routes
```

2. Install current packages

```bash
npm install --save @ddadaal/next-typed-api-routes-runtime
npm install -D @ddadaal/next-typed-api-routes-cli
```

3. Replace all imports from `@ddadaal/next-typed-api-routes` to `@ddadaal/next-typed-api-routes-runtime`. Editors and IDE can be of great help.

4. Regenerate API clients and schema

```bash
npx ntar schema && npx ntar client
```

# Roadmap

- [ ] More configurations

# License

MIT



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.3.0](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.12...v0.3.0) (2021-11-19)

**Note:** Version bump only for package @ddadaal/next-typed-api-routes





# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.12](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.11...v0.2.12) (2021-11-19)


### Bug Fixes

* necessary packages are not installed ([b333b01](https://github.com/ddadaal/next-typed-api-routes/commit/b333b01e67f7c5455488653c7e86f67fe46567f0))

### [0.2.11](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.10...v0.2.11) (2021-11-19)


### Features

* update deps and move unnecessary packages to dev ([cc244f8](https://github.com/ddadaal/next-typed-api-routes/commit/cc244f8043af6dc57e75a87bfd4d9ff9f41875cc))

### [0.2.10](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.9...v0.2.10) (2021-11-11)


### Bug Fixes

* **fetch:** handle 204 no content response ([64674a9](https://github.com/ddadaal/next-typed-api-routes/commit/64674a958f0a466cfda1177cf6c9a14597a034f8))

### [0.2.9](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.8...v0.2.9) (2021-11-11)


### Bug Fixes

* **route:** fix null response never returns ([14181ac](https://github.com/ddadaal/next-typed-api-routes/commit/14181ac70129e6cec5bbb8df4ceef921a41d66de))

### [0.2.8](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.7...v0.2.8) (2021-11-11)


### Bug Fixes

* **client:** add missing export ([09ced34](https://github.com/ddadaal/next-typed-api-routes/commit/09ced34dbbd24255b65d5af0007eab20977daed5))
* **fetch:** http handler wrong payload ([6b4e6f2](https://github.com/ddadaal/next-typed-api-routes/commit/6b4e6f27392af01632c55f65c1f85c5a294a902d))

### [0.2.7](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.6...v0.2.7) (2021-11-11)


### Features

* **fetch:** improve fetch error design ([abacc0a](https://github.com/ddadaal/next-typed-api-routes/commit/abacc0aa6247224f45fc06a3960d287ca9be6be6))

### [0.2.6](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.5...v0.2.6) (2021-11-11)

### [0.2.5](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.4...v0.2.5) (2021-11-11)

### [0.2.4](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.3...v0.2.4) (2021-11-09)


### Bug Fixes

* **cli:** api route url should not include index ([147fbb4](https://github.com/ddadaal/next-typed-api-routes/commit/147fbb4882018fa57a57f757cb7e23185f7a398f))

### [0.2.3](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.2...v0.2.3) (2021-11-09)


### Features

* **client:** improve fetch httpError type ([ea0549c](https://github.com/ddadaal/next-typed-api-routes/commit/ea0549cda452a671a13f1389cb4790e2a7f61f9b))

### [0.2.2](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.1...v0.2.2) (2021-11-09)


### Features

* improve http error handler ([0d84e87](https://github.com/ddadaal/next-typed-api-routes/commit/0d84e8762a1798a8ed044eeeb46f93107e410532))

### [0.2.1](https://github.com/ddadaal/next-typed-api-routes/compare/v0.2.0...v0.2.1) (2021-11-09)


### Features

* allow import from root package in server package ([6a2dabb](https://github.com/ddadaal/next-typed-api-routes/commit/6a2dabbfeeae22249fb561971f1a0cd337216a02))


### Bug Fixes

* **cli:** cli use wrong fetch import path ([81f0b9c](https://github.com/ddadaal/next-typed-api-routes/commit/81f0b9c845a11a3d04722acda46eb97b7405f8ef))

## [0.2.0](https://github.com/ddadaal/next-typed-api-routes/compare/v0.1.10...v0.2.0) (2021-11-06)


### ⚠ BREAKING CHANGES

* separate client and server bundle for client tree shaking

### Features

* separate client and server bundle for client tree shaking ([b41667a](https://github.com/ddadaal/next-typed-api-routes/commit/b41667a01de7234158fdca2ebdc3345c84aad407))

### [0.1.10](https://github.com/ddadaal/next-typed-api-routes/compare/v0.1.9...v0.1.10) (2021-11-04)


### Bug Fixes

* infer req.body type correctly ([e49925b](https://github.com/ddadaal/next-typed-api-routes/commit/e49925b78a8ea2b857e78951de11810417b5d6ec))

### [0.1.9](https://github.com/ddadaal/next-typed-api-routes/compare/v0.1.8...v0.1.9) (2021-11-04)


### Features

* support finally on FetchResult ([25ff1f6](https://github.com/ddadaal/next-typed-api-routes/commit/25ff1f695485f79436ac75d6cb2d14fa8e3a1d8d))


### Bug Fixes

* change coerceType to "array" ([8ce1631](https://github.com/ddadaal/next-typed-api-routes/commit/8ce1631fd9c589785db56de32b15c23e05c73d99))

### [0.1.8](https://github.com/ddadaal/next-typed-api-routes/compare/v0.1.7...v0.1.8) (2021-10-12)


### Bug Fixes

* enable coerceTypes ([045bde2](https://github.com/ddadaal/next-typed-api-routes/commit/045bde2fc2c006c5fe732abbd32715ac73d36f5e))

### [0.1.7](https://github.com/ddadaal/next-typed-api-routes/compare/v0.1.6...v0.1.7) (2021-10-12)

### [0.1.6](https://github.com/ddadaal/next-typed-api-routes/compare/v0.1.5...v0.1.6) (2021-10-12)


### Bug Fixes

* export JsonFetchResultPromiseLike ([145d068](https://github.com/ddadaal/next-typed-api-routes/commit/145d06851c7ae8ba7c8e34756c5fa47b17e7742e))

### [0.1.5](https://github.com/ddadaal/next-typed-api-routes/compare/v0.1.4...v0.1.5) (2021-09-30)


### Bug Fixes

* **package:** add long to suppress fsj warning ([f99c62a](https://github.com/ddadaal/next-typed-api-routes/commit/f99c62aca12ef6c7701cd62f59f96230cf604ec4))
* **route:** allow union types ([f7f542f](https://github.com/ddadaal/next-typed-api-routes/commit/f7f542f7f6ceacb52461b5231933deac5b9d947c))
* **route:** body validation error code ([310a315](https://github.com/ddadaal/next-typed-api-routes/commit/310a3153d238f3abff3f10da71dc79bc40a3eacb))

src/fetch/fetch.ts
import type { GeneralSchema, SuccessResponse } from "../types/schema";
import type {
  Querystring, RequestArgs,
} from "../types/request";
import { removeNullOrUndefinedKey } from "./utils";
import { failEvent, finallyEvent, prefetchEvent, successEvent } from "./events";
import { parseQueryToQuerystring } from "./parseQueryToQuerystring";
import { replacePathArgs } from "./replacePathArgs";
import { TYPE_ERROR, HttpError, FETCH_ERROR } from "./HttpError";

function isServer() {
  return typeof window === "undefined";
}

function isFormData(a: any): a is FormData {
  return !isServer() && a instanceof FormData;
}

export type HttpMethod = "GET" | "POST" | "DELETE" | "PATCH" | "PUT";

let token = "";

export function changeBearerToken(newToken: string): void {
  token = newToken;
}


export function fullFetch(
  path: string,
  query?: Querystring,
  init?: RequestInit,
): Promise<Response> {
  const headers = token
    ? { ...init?.headers, "authorization": `Bearer ${token}` }
    : init?.headers ?? {};

  let url = path;
  if (query) {
    url += parseQueryToQuerystring(query);
  }

  return fetch(url,
    {
      ...init,
      headers,
      mode: "cors",
      // disable cache for IE11
      cache: "no-cache",
    });
}

export type FullFetch = typeof fullFetch;

export interface FetchInfo {
  path: string;
  method?: HttpMethod;
  query?: Querystring;
  body?: unknown;
  headers?: Record<string, string>;
}

export type JsonFetchResult<TResp> = TResp;

function tryParseJson(text: string): object | undefined {
  try {
    return JSON.parse(text);
  } catch (e) {
    return undefined;
  }
}

function checkIsJson(resp: Response) {
  const type = resp.headers.get("content-type");
  return type && type.includes("application/json");
}

type RejectHandler<TRej> = (reason: any) => TRej | PromiseLike<TRej>;

export class JsonFetchResultPromiseLike<T extends GeneralSchema>
implements PromiseLike<SuccessResponse<T>> {

  private promise: Promise<Response>;
  private httpErrorHandler: Map<number, (error: any) => unknown>;

  constructor(
    promise: Promise<Response>,
  ) {
    this.promise = promise;
    this.httpErrorHandler = new Map();
  }

  then<TSuc = SuccessResponse<T> , TRej = unknown>(
    onfulfilled?: ((value: SuccessResponse<T> ) => TSuc | PromiseLike<TSuc>) | null,
    onrejected?: ((reason: any) => TRej | PromiseLike<TRej>) | null,
  ): Promise<TSuc | TRej> {
    return this.promise
      .then(async (resp) => {
        if (resp.ok) {
          if (resp.status === 204) {
            const obj = null as any;
            successEvent.execute({ status: 204, data: obj });
            return onfulfilled ? onfulfilled(obj) : obj;
          } else if (checkIsJson(resp)) {
            const obj = await resp.json();
            successEvent.execute({ status: resp.status, data: obj });
            return onfulfilled ? onfulfilled(obj) : obj;
          } else {
            successEvent.execute({ status: resp.status, data: resp });
            throw resp;
          }
        } else {
          const text = await resp.text();
          const payload = new HttpError(resp.status, tryParseJson(text), text);

          const handler = this.httpErrorHandler.get(resp.status);

          if (handler) {
            // if a handler is registered to the code,
            // don't fire failEvent
            const val = handler(payload.data);
            return onrejected ? onrejected(val) : val;
          } else {
            failEvent.execute(payload);
            throw payload;
          }
        }
      }).catch((r) => {

        const handleOrThrow = (payload: any) => {
          if (onrejected) {
            onrejected(payload);
          } else {
            throw payload;
          }
        };

        if (r.name === "FetchError") {
          const payload = new HttpError(FETCH_ERROR, r);
          failEvent.execute(payload);
        } else if (r instanceof TypeError) {
          const payload = new HttpError(TYPE_ERROR, r);
          failEvent.execute(payload);
        }

        // already handle or throw
        handleOrThrow(r);
      }).finally(() => {
        finallyEvent.execute(undefined);
      });
  }

  /**
   * Attach an error handler specific to a status code.
   *
   * The handler will be called when this code is received.
   * The return value of the handler will be called with onrejected.
   *
   * @param code http status code
   * @param handler handler for this type of code
   * @returns A Promise
   */
  httpError<Code extends number>(
    code: Code, handler: (err: T["responses"][Code]) => any
  ): JsonFetchResultPromiseLike<T> {
    this.httpErrorHandler.set(code, handler);
    return this as any;
  }

  catch<TRej = never>(onrejected?: RejectHandler<TRej> | null) {
    return this.then(undefined, onrejected);
  }

  finally(onfinally?: () => any): Promise<unknown> {
    return this.then(
      (v) => {
        onfinally?.();
        return v;
      },
      (r) => {
        onfinally?.();
        return r;
      });
  }

}

/**
 * Fetch and returns as json.
 * @param info the fetch info
 * @throws {Response} If the request is successful but response is not json,
 * the response will be thrown
 * @throws {JsonFetchError} If the statusCode is not [200, 300), a error will be thrown
 */
export function jsonFetch<T extends GeneralSchema>(
  info: FetchInfo,
  signal?: AbortSignal,
): JsonFetchResultPromiseLike<T> {

  const isForm = isFormData(info.body);

  prefetchEvent.execute(undefined);

  return new JsonFetchResultPromiseLike(fullFetch(info.path, info.query, {
    method: info.method ?? "GET",
    headers: {
      ...isForm ? undefined : { "content-type": "application/json" },
      ...info.headers,
    },
    body: isForm ? (info.body as any) : JSON.stringify(info.body),
    signal,
  }));
}

export type JsonFetch = typeof jsonFetch;

export function fromApi<TSchema extends GeneralSchema>(method: HttpMethod, url: string) {
  return function (
    args: RequestArgs<TSchema>,
    signal?: AbortSignal,
  ): JsonFetchResultPromiseLike<TSchema>  {

    const anyArgs = args as any;
    // replace path params using query
    const replacedPath = anyArgs.query
      ? replacePathArgs(url, anyArgs.query)
      : url;

    return jsonFetch({
      path: replacedPath,
      method: method,
      query: removeNullOrUndefinedKey(anyArgs.query),
      body: anyArgs.body,
    }, signal);
  };
}
